In [5]:
import subprocess as sub
import os

import matplotlib.pyplot as plt
import math

In [3]:
K_VALUES =      [10 ** i for i in range(3)]
N_MIN_VALUES =  [i for i in range(2, 10)]
N_AVG_VALUES =  [10 * i for i in range(2, 21)]
N_MAX_VALUES =  [1000 * i for i in range(2, 21)]
ALGORITHMS =    ['insertion_sort', 'merge_sort', 'quick_sort', 'dual_pivot_quick_sort', 'hybrid_sort']

print(K_VALUES)
print(N_MIN_VALUES)
print(N_AVG_VALUES)
print(N_MAX_VALUES)
print(ALGORITHMS)

[1, 10, 100]
[2, 3, 4, 5, 6, 7, 8, 9]
[20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]
[2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 11000, 12000, 13000, 14000, 15000, 16000, 17000, 18000, 19000, 20000]
['insertion_sort', 'merge_sort', 'quick_sort', 'dual_pivot_quick_sort', 'hybrid_sort']


In [8]:
def test_algorithm(algorithm, data_type, repeats, n_values):
    comparisons_data = []
    swaps_data = []
    for n in n_values:
        comparisons_sum = 0
        swaps_sum = 0
        for k in range(repeats):
            out = sub.run('./data_generator {} {} | ./{}'.format(data_type, n, algorithm), 
                            shell=True, capture_output=True, encoding="utf-8")
            result = out.stdout.split('\n')

            if len(result) < 3:
                continue
            
            success = bool(result[0])
            comparisons = int(result[1])
            swaps = int(result[2])

            comparisons_sum += comparisons
            swaps_sum += swaps

        comparisons_data.append(comparisons_sum / repeats / n / math.log(n))
        swaps_data.append(swaps_sum / repeats / n / math.log(n))

    return comparisons_data, swaps_data
            

In [518]:
data = {}
for k in K_VALUES:
    data[k] = {}
    for alg in ALGORITHMS:
        data[k][alg] = test_algorithm(alg, 'r', k, N_MAX_VALUES)


In [2]:
colors = ('r', 'g', 'b', 'y', 'm')

def compare_algorithms(x, data):
    for k, series in data.items():
        fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(9, 6))
        fig.suptitle('Algorithms Comparison (random data;  k = ' + str(k) + ';  {} < n < {})'.format(x[0], x[-1]))
        ax1.set_xlabel('n')
        ax1.set_title('Comparision Number')
        ax2.set_xlabel('n')
        ax2.set_title('Swaps Number')

        labels = []
        lines = []

        i = 0
        for alg, ys in series.items():
            l1 = ax1.plot(x, ys[0], color=colors[i])
            l2 = ax2.plot(x, ys[1], color=colors[i])

            lines.append(l1)   
            labels.append(alg.replace('_', ' ').title())  

            i += 1     

        fig.legend(lines, labels=labels, loc='upper right', bbox_to_anchor = (1.16, 0.6))

        print('charts/comparision_k={}_n=[{},{}]'.format(k, x[0], x[-1]))
        plt.plot()

In [ ]:
compare_algorithms(N_MAX_VALUES, data)

In [13]:
for i in range(1000) :
    out = sub.run('./data_generator r 1000 | ./dual_pivot_quick_sort', shell=True, capture_output=True, encoding="utf-8")
    result = out.stdout.split('\n')

    success = bool(result[0])

    if not success:
        print('NOT SUCCESS')

            

In [11]:
quick_sort_data = test_algorithm('quick_sort', 'r', 100, N_MAX_VALUES)
dual_pivot_quick_sort_data = test_algorithm('dual_pivot_quick_sort', 'r', 100, N_MAX_VALUES)

In [12]:
C1 = sum(quick_sort_data[0]) / len(quick_sort_data[0])
C2 = sum(quick_sort_data[1]) / len(quick_sort_data[1])

print(C1)
print(C2)

D1 = sum(dual_pivot_quick_sort_data[0]) / len(dual_pivot_quick_sort_data[0])
D2 = sum(dual_pivot_quick_sort_data[1]) / len(dual_pivot_quick_sort_data[1])

print(D1)
print(D2)

2.064236666274528
0.3461860526410561
1.5454639372851018
0.5223202202832493


Wyznaczanie stalej przy nlogn dla QuickSort oraz DualPivot QuickSort:

Quick Sort: 

    liczba porownan : 2.06*nlog(n)

    liczba swap'ow  : 0.35*n log(n)

Dual Pivot Quick Sort: 
    
    liczba porownan : 1.55*nlog(n)
    
    liczba swap'ow  : 0.5*nlog(n)